addons/website_event_booth_exhibitor/models/event_booth.py

```diff
@@ -11,42 +11,38 @@ class EventBooth(models.Model):
     sponsor_type_id = fields.Many2one(related='booth_category_id.sponsor_type_id')
     sponsor_id = fields.Many2one('event.sponsor', string='Sponsor', copy=False)
 
-    def write(self, vals):
-        for booth in self:
-            if vals.get('partner_id') and booth.use_sponsor and not booth.sponsor_id and not vals.get('sponsor_id'):
-                vals['sponsor_id'] = booth._get_sponsor(vals)
-        return super(EventBooth, self).write(vals)
-
     def action_view_sponsor(self):
         action = self.env['ir.actions.act_window']._for_xml_id('website_event_exhibitor.event_sponsor_action')
         action['views'] = [(False, 'form')]
         action['res_id'] = self.sponsor_id.id
         return action
 
-    def _get_sponsor(self, vals):
+    def _get_or_create_sponsor(self, vals):
         self.ensure_one()
         sponsor_id = self.env['event.sponsor'].sudo().search([
-            ('partner_id', '=', vals.get('partner_id')),
+            ('partner_id', '=', self.partner_id.id),
             ('sponsor_type_id', '=', self.sponsor_type_id.id),
             ('exhibitor_type', '=', self.booth_category_id.exhibitor_type),
             ('event_id', '=', self.event_id.id),
         ], limit=1)
         if not sponsor_id:
+            len_sponsor_key = len('sponsor_')
             values = {
-                'event_id': vals.get('event_id') or self.event_id.id,
+                'event_id': self.event_id.id,
                 'sponsor_type_id': self.sponsor_type_id.id,
                 'exhibitor_type': self.booth_category_id.exhibitor_type,
-                'partner_id': vals.get('partner_id'),
+                'partner_id': self.partner_id.id,
+                **{key[len_sponsor_key:]: value for key, value in vals.items() if key.startswith('sponsor_')},
```

Comment:
You may use standard string split tools like `partition`: ``key.partition("sponsor_")[2]: value for key, value ... `` Partition splits on the first found occurence, like
```
In [4]: a = "sponsor_name"

In [5]: a.partition("sponsor_")
Out[5]: ('', 'sponsor_', 'name')

In [6]: a.partition("sponsor_")[2]
Out[6]: 'name'
```